crates/ruff_python_formatter/src/expression/expr_slice.rs

```diff
@@ -1,23 +1,257 @@
+        );
+        debug_assert!(
+            !(step.is_some() && second_colon_partition_index < slice_dangling_comments.len())
+        );
+
+        // Handle spacing around the colon(s)
+        // https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html#slices
+        let lower_simple = lower.as_ref().map_or(true, |expr| is_simple_expr(expr));
+        let upper_simple = upper.as_ref().map_or(true, |expr| is_simple_expr(expr));
+        let step_simple = step.as_ref().map_or(true, |expr| is_simple_expr(expr));
+        let all_simple = lower_simple && upper_simple && step_simple;
+
+        // lower
+        if let Some(lower) = lower {
+            lower.format().fmt(f)?;
+            let lower_trailing_comments = comments.trailing_comments(lower.as_ref().into());
+            if !lower_trailing_comments.is_empty() {
+                // make sure the colon is after the comments
+                hard_line_break().fmt(f)?;
+            }
+        } else {
+            dangling_comments(&slice_dangling_comments[..first_colon_partition_index]).fmt(f)?;
+        }
+
+        // First colon
+        // The spacing after the colon depends on both the lhs and the rhs:
+        // ```
+        // e00 = x[:]
+        // e01 = x[:1]
+        // e02 = x[: a()]
+        // e10 = x[1:]
+        // e11 = x[1:1]
+        // e12 = x[1 : a()]
+        // e20 = x[a() :]
+        // e21 = x[a() : 1]
+        // e22 = x[a() : a()]
+        // e200 = "e"[a() : :]
+        // e201 = "e"[a() :: 1]
+        // e202 = "e"[a() :: a()]
+        // ```
+        if !all_simple {
+            space().fmt(f)?;
+        }
+        text(":").fmt(f)?;
+        // No upper node, no need for a space, e.g. `x[a() :]`
+        if !all_simple && upper.is_some() {
+            space().fmt(f)?;
+        }
+
+        // Upper
+        if let Some(upper) = upper {
+            let upper_leading_comments = comments.leading_comments(upper.as_ref().into());
+            if !upper_leading_comments.is_empty() {
+                // We can likely do this better by putting two spaces here if the comment was
+                // on the same line as the colon in the source, but this is not a bad formatting
+                // for now either
+                hard_line_break().fmt(f)?;
+            }
+            upper.format().fmt(f)?;
+            let upper_trailing_comments = comments.trailing_comments(upper.as_ref().into());
+            if !upper_trailing_comments.is_empty() {
+                // make sure the colon is after the comments
+                hard_line_break().fmt(f)?;
+            }
+        } else {
+            let dangling =
+                &slice_dangling_comments[first_colon_partition_index..second_colon_partition_index];
+            if !dangling.is_empty() {
+                // put the colon and comments on their own lines
+                hard_line_break().fmt(f)?;
+            }
+            dangling_comments(dangling).fmt(f)?;
```

Comment:
Nit: We shouldn't call into the dangling comments formatting if we know that there are none. 

```suggestion
            if !dangling.is_empty() {
                // put the colon and comments on their own lines
                write!(f, [hard_line_break(), dangling_comments(dangling)])?;
            }
```